sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala

```diff
@@ -39,27 +40,49 @@ import org.apache.spark.unsafe.types.UTF8String
  * @param requiredSchema The schema of the data that should be output for each row. This should be a
  *                       subset of the columns in dataSchema.
  * @param options Configuration options for a CSV parser.
+ * @param filters The pushdown filters that should be applied to converted values.
  */
 class UnivocityParser(
     dataSchema: StructType,
     requiredSchema: StructType,
-    val options: CSVOptions) extends Logging {
+    val options: CSVOptions,
+    filters: Seq[Filter]) extends Logging {
   require(requiredSchema.toSet.subsetOf(dataSchema.toSet),
     s"requiredSchema (${requiredSchema.catalogString}) should be the subset of " +
       s"dataSchema (${dataSchema.catalogString}).")
 
+  def this(dataSchema: StructType, requiredSchema: StructType, options: CSVOptions) = {
+    this(dataSchema, requiredSchema, options, Seq.empty)
+  }
   def this(schema: StructType, options: CSVOptions) = this(schema, schema, options)
 
   // A `ValueConverter` is responsible for converting the given value to a desired type.
   private type ValueConverter = String => Any
 
+  private val csvFilters = new CSVFilters(
+    filters,
+    dataSchema,
+    requiredSchema,
+    options.columnPruning)
+
+  private[sql] val parsedSchema = csvFilters.readSchema
+
+  // Mapping of field indexes of `parsedSchema` to indexes of `requiredSchema`.
+  // It returns -1 if `requiredSchema` doesn't contain a field from `parsedSchema`.
```

Comment:
when can this happen?

Reply:
For example:
dataSchema: `i INTEGER, d DOUBLE, s STRING`
requiredSchema: `i INTEGER`
filters: `Seq(LessThan("d", 10))`
parsedSchema: `i INTEGER, d DOUBLE`
In that case, `requiredSchema` doesn't contain a field from `parsedSchema`

Reply:
Is that case possible? If a filter references a column, it should be in `requiredSchema`.

Reply:
What should the upper layer do with the column if a datasource already applied filters? As far as I know filters are applied only once in DSv2, @cloud-fan or I am wrong?

Reply:
It should read the column referenced by filters from the source because we filter it one more time within Spark side.

Reply:
If `requiredSchema` always contains filter references, it is significant assumption, and it can simplify this implementation slightly. Is it just specific of current implementation or `requiredSchema` could contain **really required** column in the future? because filter columns are not actually required if the filter is applied only once in a datasource.

Reply:
> Is it just specific of current implementation or `requiredSchema` could contain **really required** column in the future?

We will still have to do the filter at Spark side for some other datasources where record-level filter isn't supported. As an example, ORC datasource.

In fact, Parquet does not do record-level filter by default because filtering out in Spark side was found faster.

So, I can quite clearly foresee the `requiredSchema` will contain the column referenced by pushed filters, or at the very least we will keep one way to do it.